Read and write support for pixel values in GDALBands

yellowcap opened the pull request:
- Allow instantiation of GDALRasters from json/dict data
- Added read and write support for GDALBand pixel values throught `data` attribute.
- Added setter functions for some other `GDALRaster` and `GDALBand` properties: `srs`, `geotransform`, `nodata_value`.
- Support for `GDALFlushCache` in gdal C prototypes
- Added private flush method to GDALRaster to make sure all
  data is written to files when file-based rasters are changed.
- Replaced `.ptr` with `._ptr` for internal pointer attribute

The idea behind the instantiation of GDALRasters from ``json`` or ``dict``  data is to have a flexible way of converting DB values into GDALRaster instances.

PostGIS Rasters, SpatialLite Rasters and Oracle GeoRasters are all defined quite differently and do not follow a common standard. This is why the instanciation of GDALRasters needs to be flexible, the idea is to have the DB wrappers/adapters to convert the differnt output from the databases into a common format: a dictionary with the parsed data.


Refs #23804.

yellowcap commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -57,6 +70,11 @@ def __init__(self, ds_input, write=False):
                         ds_input, err))
             else:
                 raise GDALException('Unable to read raster source input "{}"'.format(ds_input))
+        elif isinstance(ds_input, dict):
```
This I am least sure about, I was trying to keep the input parameter to a single object, but baybe this should be changed to kwargs? The challenge is to "tell" the function if the input is a path to a file or a set of parameters, which is an either/or case.

claudep commented:
Daniel, could you please make a separate pull request for each commit? I'd like to review them separately, even if it's nice to see RasterField working based on the GDALRaster write support.

yellowcap commented:
I restructured this pull request and removed the RasterField part. See the comments in the pull request description regarding GDALRaster instantiation, I am not sure if the instantiation is pythonic enough, but its something along those lines that will be necessary for the RasterField.

claudep commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -67,3 +66,73 @@ def nodata_value(self):
+                return numpy.frombuffer(
+                    data_array, dtype=numpy.dtype(data_array)).reshape(size)
+            else:
+                return list(data_array)
```
At first, I was a bit surprised to have the same method for both reading and writing. I recognize that having `get_data`/`set_data` would sound Java-ish, but could maybe convey a bit more the idea of a possibly costly operation... Thoughts?

yellowcap replied:
I understand the concern, retrieving pixel values of a big raster fills up the memory quickly.

One simple option to aleviate this a bit (aside making two functions), is to warn about this in the documentation with a note/warning banner or so. The other option is to make the size argument a requirement, so that the user is very aware of the number of pixels he or she is retrieving.

I lean towards keeping just one function, but I am open to both solutions. A single function mirrors the underlying gdal function and splitting it into two would probably require making the dual one private and write two "wrappers" for get and set.

yellowcap commented:
Regarding the imports, I had corrected that in the second commit, as the build checks gave me errors. I moved the ordering of the imports to the first commit where they belong. Thanks for pointing that out.

claudep commented:
I've pushed the numpy patch, you should be able to rebase.

claudep commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -67,3 +66,76 @@ def nodata_value(self):
+    def data(self, data=None, offset=None, size=None, as_buffer=False):
+        """
+        Reads or writes pixel values for this band. Blocks of data can
+        be accessed by specifying the with and height and the offset of
```
with -> width

claudep commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -67,3 +66,76 @@ def nodata_value(self):
+            raise ValueError('Size is larger than raster.')
+
+        # Create ctypes type array generator
+        ctypes_array = GDAL_TO_CTYPES[self.datatype()] * (size[0]*size[1])
```
flake8 complains about missing spaces around `*`

claudep commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -67,3 +66,76 @@ def nodata_value(self):
+            # Set write mode
+            access_flag = 1
+            # Instantiate ctypes array holding the input data
+            if isinstance(data, (str, buffer)):
```
`str` is different between Python 2 and 3, what type do you target exactly? Should it be `bytes`?

claudep commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -67,3 +66,76 @@ def nodata_value(self):
+        # Create ctypes type array generator
+        ctypes_array = GDAL_TO_CTYPES[self.datatype()] * (size[0]*size[1])
+
+        if data is not None:
```
I'd reverse this to `if data is None`, to match the structure below (and prevent unnecessary negation).

claudep commented on django/contrib/gis/gdal/raster/const.py:
```diff
@@ -17,3 +20,6 @@
 }
+
+GDAL_TO_CTYPES = [None, c_byte, c_uint16, c_int16, c_uint32, c_int32,
+                  c_float, c_double, None, None, None, None]
```
I'd add a comment about the `GDAL_TO_CTYPES` index meaning (should match datatype I guess)

claudep commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,23 +61,68 @@ def __init__(self, ds_input, write=False):
+            ds_input = json.loads(ds_input)
+
         # If input is a valid file path, try setting file as source.
         if isinstance(ds_input, six.string_types):
```
`elif` might be clearer (I understand it's not necessary)

yellowcap replied:
I tried to use elif here, but the problem is that json inputs need some kind of pre-pocessing. These lines should check if the input string is a json. If yes, convert the input to a dict - and subsequently still use the dict option for instantiation.

claudep replied:
Oh, got it. Then a comment would be nice (for stupid readers like me)

claudep commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,23 +61,68 @@ def __init__(self, ds_input, write=False):
+                raise GDALException('Specify name for creation of raster '\
+                                    'with driver "{}".'.format(driver.name))
+
+            # Check if width and height where specified
```
where -> were

claudep commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,23 +61,68 @@ def __init__(self, ds_input, write=False):
+
+            # Set band data if provided
+            for i in range(len(ds_input.get('bands', []))):
+                bnd_input = ds_input['bands'][i]
```
`band_input`, you don't get much by saving one char :-)

claudep commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,23 +61,68 @@ def __init__(self, ds_input, write=False):
+            self.srs = ds_input.get('srid', 0)
+
+            # Set additional properties if provided
+            if ds_input.has_key('origin'):
```
`has_key` is deprecated, use `if 'origin' in ds_input:` instead. Also please replace other occurrences in the patch.

claudep commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,23 +61,68 @@ def __init__(self, ds_input, write=False):
+                self.skew.x, self.skew.y = ds_input['skew']
         else:
-            raise GDALException('Invalid data source input type: "{}".'.format(type(ds_input)))
+            raise GDALException('Invalid data source input type: '\
```
We try to avoid line continuation backslashes in Django, and it's not a problem to overcome the 80-chars limit when it makes readability better.

claudep commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,23 +61,68 @@ def __init__(self, ds_input, write=False):
-        if self._ptr and capi:
-            capi.close_ds(self._ptr)
+        if self.ptr and capi:
+            capi.close_ds(self.ptr)
```
We might instead replace other instances by `self._ptr`, to convey the idea of a private and internal variable. Just a suggestion.

claudep commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -74,12 +133,28 @@ def __repr__(self):
 
+    def _flush(self):
+        """
+        Flushes all data from memory into the source file if it exists.
```
We are now aiming for the PEP 257 convention, that is using `Flush` instead of `Flushes`.

claudep commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -41,6 +41,7 @@ determine where these sample files are installed on your computer::
     >>> import django.contrib.gis
     >>> GIS_PATH = os.path.dirname(django.contrib.gis.__file__)
     >>> CITIES_PATH = os.path.join(GIS_PATH, 'tests/data/cities/cities.shp')
+    >>> RASTER_PATH = os.path.join(GIS_PATH, 'gdal/tests/data/raster.tif')
```
With the recent move of tests of contrib apps, these files are not distributed any longer with Django. This is not directly related to this patch, but we might point to Github instead.

claudep commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1100,35 +1101,95 @@ one or more layers of data named bands.  Each band, represented by a
+
+.. class:: GDALRaster(ds_input, write=False)
+
+    The constructor for ``GDALRaster`` accepts a two parameters. The first parameter
```
Remove `a`

claudep commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1100,35 +1101,95 @@ one or more layers of data named bands.  Each band, represented by a
+        The name of the GDAL driver used to handle the input file. For rasters created
+        from a file, the Driver type is detected automatically. The creation of rasters from
+        scratch is a in-memory raster by default (``'MEM'``), but can be altered as
+        needed. An in-memory raster is created through the following example:
```
Doesn't play well with the next line: `in-memory` - `GeoTiff` ??

claudep commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1100,35 +1101,95 @@ one or more layers of data named bands.  Each band, represented by a
     .. attribute:: height
 
-        The height of the source in pixels (Y-axis).
+        The height of the source in pixels (Y-axis).:
```
`.` or `:`, not both.

claudep commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1192,6 +1301,10 @@ blue.
 
+    .. attribute:: pixel_count
+
+      The total number of pixels in this band. Is equal to ``with*height``.
```
width

claudep commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1215,6 +1328,51 @@ blue.
+      be requested by specifying an offset and block size.
+
+      If NumPy is available, the data is returned as NumPy array. For large rasters
+      its is highly recommended to use NumPy for performance reasons.
```
`it is`

claudep commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1215,6 +1328,51 @@ blue.
+      Example:
+
+        >>> rst = GDALRaster({'width': 4, 'height': 4, 'datatype': 1})
+        >>> bnd =rst.bands[0]
```
Missing space

claudep commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -41,9 +41,13 @@
 import unittest
 
 from django.contrib.gis.gdal import HAS_GDAL
+from django.contrib.gis.gdal.tests.data.textrasters import JSON_RASTER
```
The path needs updating (I'd use a relative one)

claudep commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -51,7 +55,6 @@
     from django.contrib.gis.gdal import GDALRaster
     from django.contrib.gis.gdal.raster.band import GDALBand
 
-
```
PEP 8 requires this blank line

claudep commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -61,6 +64,7 @@ def setUp(self):
         self.rs_path = os.path.join(os.path.dirname(upath(__file__)),
                                     'data/raster.tif')
         self.rs = GDALRaster(self.rs_path)
+        self.rsmem = GDALRaster(JSON_RASTER)
```
I'm not sure if it's a good idea to define it in `setUp`, let's define it in the method where it is used.

claudep commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -90,14 +94,58 @@ def test_geotransform_and_friends(self):
 
+    def test_file_based_raster_creation(self):
+        # Prepare tempfile path
+        filename = os.path.join(tempfile.gettempdir(), 'file_based_raster.tif')
```
Might be better to use a `tempfile.NamedTemporaryFile` here, maybe with `delete=False` if it is closed before the end of the test. Then, we can call `self.addCleanup(...)` to ensure the file is deleted at the end of the test.

claudep commented:
Nice work! We'll still have to ask for a language review by a native speaker before the commit.

yellowcap commented:
Thanks for all the comments, I will integrate your suggestions next week.

In the meantime, I made some addions to the django-raster package to make it more similar to the django code here.

So it now basically has become a testbed for the work around the RasterField we are targeting here. It works really well, have a look here if you want to see django rendering tiles (this is still relying on the  python-gdal bindings, but our django version will perform similarly)

http://sfwsc.geodesigntech.com/scenarios/

claudep commented:
This is really exciting. It's very precious to be able to shape this functionality around real-world examples, Also precious to collaborate with people ready to give back to the community a part of their genius!

yellowcap commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -72,15 +134,35 @@ def __repr__(self):
+        # Raise an Exception if the value is being changed in read mode.
+        if not self._write:
+            raise GDALException(
+                'Raster needs to be opened in write mode to change values.')
```
Added an exception here, I realized that GDAL does not throw errors when changes are made on files opened in read mode, so this has to be checked on the python side.

timgraham replied:
single line?

yellowcap commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -33,14 +33,18 @@ Sample Data
+    ...     urllib.urlretrieve(base_url + 'cities/' + name, os.path.join(destination + name))
+    >>> urllib.urlretrieve(base_url + 'rasters/raster.tif', os.path.join(destination + 'raster.tif'))
+    >>> CITIES_PATH = os.path.join(destination, 'cities.shp')
+    >>> RASTER_PATH = os.path.join(destination, 'raster.tif')
```
This is not elegant at all, but should work. Not sure if that was along the lines of what you were thinking to do.

claudep replied:
Not convinced either. I'll fix that in a separate commit by simply giving a github link to the downloadable file(s).

claudep replied:
Done here: https://github.com/django/django/commit/1825365152c6474255f976dca4a68a8e3f070afa

yellowcap commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -72,15 +134,35 @@ def __repr__(self):
+        Returns the name of this raster. Corresponds to filename
+        for file-based rasters.
+        """
+        return force_text(capi.get_ds_description(self._ptr))
```
I replaced the `ptr` with `_ptr` as suggested.

yellowcap commented:
I updated the pull request, adopting all comments, most of it was straightforward. I made some inline comments for some of the more interesting changes.

yellowcap commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -59,7 +65,7 @@ class GDALRasterTests(unittest.TestCase):
     def setUp(self):
         self.rs_path = os.path.join(os.path.dirname(upath(__file__)),
-                                    'data/raster.tif')
+                                    '../data/rasters/raster.tif')
```
Moved the test data into the gis_tests data folder.

yellowcap commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
+      >>> from django.contrib.gis.gdal.raster.source import GDALRaster
+      >>> rst = GDALRaster('/path/to/your/raster.tif', write=False)
+      >>> rst.name
+      '/path/to/your/raster.tif'
```
Adopted this to new sample data retrieval. Also added crossref above to new link for the raster file download in Sample Data section.

claudep commented:
buildbot, test this please.

claudep commented:
Flake8 failures:
```
./django/contrib/gis/gdal/raster/source.py:88:41: E713 test for membership should be 'not in'
./django/contrib/gis/gdal/raster/source.py:93:16: E713 test for membership should be 'not in'
./django/contrib/gis/gdal/raster/source.py:230:26: E226 missing whitespace around arithmetic operator
./tests/gis_tests/gdal_tests/test_raster.py:138:21: E241 multiple spaces after ':'
./tests/gis_tests/gdal_tests/test_raster.py:139:20: E241 multiple spaces after ':'
./tests/gis_tests/gdal_tests/test_raster.py:141:63: E226 missing whitespace around arithmetic operator
./tests/gis_tests/gdal_tests/test_raster.py:213:54: E231 missing whitespace after ','
./tests/gis_tests/gdal_tests/test_raster.py:213:56: E231 missing whitespace after ','
./tests/gis_tests/gdal_tests/test_raster.py:213:58: E231 missing whitespace after ','
./tests/gis_tests/gdal_tests/test_raster.py:233:51: W291 trailing whitespace
```

yellowcap commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -65,5 +64,85 @@ def nodata_value(self):
+            # working in both Python 2 and 3 (the use of the memoryview
+            # directly fails in python 2).
+            if isinstance(data, memoryview):
+                data = data.tobytes()
```
I removed the use of `buffer` to make this compatible with Python 3. I ran into a small problem with the  `from_buffer_copy` ctypes method, in 2.7 this method throws an error when passing a memoryview. As a workaround, I convert the data to bytes first, then it is passed to ctypes as bytes.

claudep replied:
We have a small compatibility layer in `django.utils.six` (https://github.com/django/django/blob/master/django/utils/six.py#L843). Could you see if it helps you?

yellowcap commented:
Flake8 failures corrected.

yellowcap commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -65,5 +65,80 @@ def nodata_value(self):
+            access_flag = 1
+
+            # Instantiate ctypes array holding the input data
+            if isinstance(data, (bytes, six.memoryview, numpy.ndarray)):
```
Yes, that helped. This works now (provided that in Python 2.7 the input is six.memoryview and not the "normal" memoryview).

yellowcap commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -116,3 +167,85 @@ def test_band_data(self):
+            self.assertEqual(bandmem.data(offset=(1, 1), size=(2, 2)), [1, 2, 3, 4])
+
+        # Set data from memoryview
+        bandmem.data(six.memoryview(struct.pack('<' + 'B B B B', *[1, 2, 3, 4])), (1, 1), (2, 2))
```
Hope this is the right way of using six...

claudep commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
+
+          >>> import tempfile
+          >>> rstfile = tempfile.NamedTemporaryFile(suffix='.tif')
+          >>> rst = GDALRaster({'driver': 'GTiff', 'name': rstfile.name, 
```
There is a trailing space here (git warned me about it).

claudep commented:
buildbot, test this please.

claudep commented:
I think it's now time time to write some release notes.

yellowcap commented:
How about:

    * The GDAL interface now supports instantiating file-based and in-memory
      :ref:`GDALRaster objects <raster-data-source-objects>` from raw data.
      Setters for raster properties such as projection or pixel values have
      been added. Numpy is used to handle pixel values if possible.

claudep commented:
Looks good. Not sure the last sentence about numpy is really useful in release notes.

yellowcap commented on docs/releases/1.9.txt:
```diff
@@ -54,7 +54,11 @@ Minor features
+  Setters for raster properties such as projection or pixel values have
+  been added.
+
 
```
Releasenotes added.

claudep commented:
@timgraham may I kindly ask you to put this PR on your review stack (especially language-wise)?

timgraham commented:
buildbot, add to whitelist.

timgraham commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -65,5 +65,80 @@ def nodata_value(self):
+    def data(self, data=None, offset=None, size=None, as_memoryview=False):
+        """
+        Reads or writes pixel values for this band. Blocks of data can
+        be accessed by specifying the width and height and the offset of
```
width, height, and offset

timgraham commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -65,5 +65,80 @@ def nodata_value(self):
+        the desired block. The same specification can be used to update
+        parts of a raster by providing an array of values.
+
+        Allowed input data types are bytes, memoryview, list, tuple and array.
```
comma after tuple

timgraham commented on django/contrib/gis/gdal/raster/band.py:
```diff
@@ -65,5 +65,80 @@ def nodata_value(self):
+        if not size:
+            size = (self.width - offset[0], self.height - offset[1])
+
+        if any((x <= 0 for x in size)):
```
extra parens needed?

timgraham commented on django/contrib/gis/gdal/raster/const.py:
```diff
@@ -17,3 +20,10 @@
     11: 'GDT_CFloat64',  # Complex Float64
 }
+
+# Lookup values to convert GDAL Pixel Type indices into ctypes objects.
```
"pixel type" (caps needed?)

timgraham commented on django/contrib/gis/gdal/raster/const.py:
```diff
@@ -17,3 +20,10 @@
+
+# Lookup values to convert GDAL Pixel Type indices into ctypes objects.
+# The GDAL band-io works with ctypes arrays to hold data to be written
+# or hold the space for data to be read into. The lookup below helps
```
or to hold

timgraham commented on django/contrib/gis/gdal/raster/const.py:
```diff
@@ -17,3 +20,10 @@
+# The GDAL band-io works with ctypes arrays to hold data to be written
+# or hold the space for data to be read into. The lookup below helps
+# selecting the right ctypes object for a given gdal pixel type.
+GDAL_TO_CTYPES = [None, c_byte, c_uint16, c_int16, c_uint32, c_int32,
```
prefer hanging indent style:
```
GDAL_TO_CTYPES = [
    None, ...
    ....
]
```

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,19 +61,68 @@ def __init__(self, ds_input, write=False):
+                self._ptr = capi.open_ds(force_bytes(ds_input), self._write)
+            except GDALException as err:
+                raise GDALException(
+                    'Could not open the datasource at "{}" ({}).'.format(ds_input, err))
```
I prefer putting the closing ) on the next line

timgraham replied:
Actually I would probably keep most of those exceptions as a single line -- we prefer long erlines when it improves readability.

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,19 +61,68 @@ def __init__(self, ds_input, write=False):
+            # Check if width and height where specified
+            if 'width' not in ds_input or 'height' not in ds_input:
+                raise GDALException(
+                    'Specify width and height attributes for JSON or Dict input.')
```
dict

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,19 +61,68 @@ def __init__(self, ds_input, write=False):
+                    'Specify width and height attributes for JSON or Dict input.')
+
+            # Create GDAL Raster
+            self._ptr = capi.create_ds(driver._ptr,
```
prefer hanging indent style with 1 arg per line

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,19 +61,68 @@ def __init__(self, ds_input, write=False):
+                                      ds_input.get('datatype', 6), None)
+
+            # Set band data if provided
+            for i in range(len(ds_input.get('bands', []))):
```
I think `enumerate` would work here

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -103,33 +186,71 @@ def srs(self):
+    @srs.setter
+    def srs(self, value):
+        """
+        Sets the Spatial Reference used in this GDALRaster. The input can be
```
no space for "Spatial Reference"? (or no caps if it's not meant to refer to the class)

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -103,33 +186,71 @@ def srs(self):
+        if sum([isinstance(x, (int, float)) for x in value]) != 6:
+            raise ValueError('Geotransform must consist of 6 numeric values.')
+        # Create ctypes double array with input and write data
+        value = (c_double * 6)(value[0], value[1], value[2],
```
single line?

yellowcap replied:
Would `value = (c_double * 6)(*value)` be ok to use here, or is the explicit version better?

claudep replied:
For me, `*value` is fine. But should we name the parameter `values`?

yellowcap replied:
We have a way of using names to change the geotransform using the `origin`, `scale` and `skew` attributes through the `TransformPoint` objects. I think if we want to name the values here, we would have to change that logic too, as it relies on the array and the corresponding positions. Currently the geotransform array is a property. But you might be right, naming them would be much easier to use when working with the geotransform attribute. I am open to both, what do you think?

claudep replied:
Oh, I just meant to suggest the `value` to `values` (plural) renaming, not more...

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -27,6 +27,8 @@ __ http://www.gdal.org/ogr/
 Overview
 ========
 
+.. _sample_data:
```
try to use a less generic title to prevent conflicts with other docs, e.g. gdal_sample_data

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
-    The constructor for ``GDALRaster`` accepts a single parameter: the path of
-    the file you want to read.
+     For raster data there is no difference between a raster instance and its
+     data source. Unlike for the Geometries objects, :class:`GDALRaster` objects are
```
Geometry objects

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
+     For raster data there is no difference between a raster instance and its
+     data source. Unlike for the Geometries objects, :class:`GDALRaster` objects are
+     always a data source. Temporary rasters can be instantiated in memory
+     using the corresponding driver, but will be of the same class as file-based
```
they will be

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
-        ``GTiff`` for a ``GeoTiff`` file. See also the `GDAL Raster Formats`__
-        list.
+        The name of the GDAL driver used to handle the input file. For GDALRasters created
+        from a file, the Driver type is detected automatically. The creation of rasters from
```
driver (or add backticks if it's meant to refer to the class)

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
-        The name of the GDAL driver used to handle the input file. For example,
-        ``GTiff`` for a ``GeoTiff`` file. See also the `GDAL Raster Formats`__
-        list.
+        The name of the GDAL driver used to handle the input file. For GDALRasters created
```
```
``GDALRasters``\s
```

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
+
+.. class:: GDALRaster(ds_input, write=False)
+
+    The constructor for ``GDALRaster`` accepts two parameters. The first parameter
```
Please use 4 space indent.

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
 
+        An in-memory raster is created through the following example:
+
+          >>> GDALRaster({'width': 10, 'height': 10}).driver.name
```
4 space indent

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1096,35 +1099,101 @@ one or more layers of data named bands.  Each band, represented by a
-        :class:`SpatialReference` instance.
+        The spatial reference system of the `source`, as a
+        :class:`SpatialReference` instance. The SRS can be changed by
+        setting it to an other SpatialReference or providing any input
```
backticks for SpatialReference

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1211,6 +1326,48 @@ blue.
+    .. method:: data(data=None, offset=None, size=None)
+
+      The accessor to the pixel values of the ``GDALBand``. Returns the complete
+      data array if no parameters are provided, a subset of the pixel array can
```
comma -> semicolon (or separate sentence)

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1211,6 +1326,48 @@ blue.
+      data array if no parameters are provided, a subset of the pixel array can
+      be requested by specifying an offset and block size as tuples.
+
+      If NumPy is available, the data is returned as NumPy array. It is highly
```
as a NumPy
For performance reasons, it is highly

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1211,6 +1326,48 @@ blue.
+
+      Data is written to the ``GDALBand`` if the ``data`` parameter is provided.
+      The input can be of one of the following types - packed string, buffer, list,
+      array and NumPy array. The number of items in the input must correspond to the
```
comma after array

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -54,7 +54,11 @@ Minor features
+  :ref:`GDALRaster objects <raster-data-source-objects>` from raw data.
+  Setters for raster properties such as projection or pixel values have
+  been added.
+
```
chop extra newline

timgraham commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -116,3 +167,85 @@ def test_band_data(self):
+        # Set data for entire dataset
+        bandmem.data(range(100))
+        if numpy:
+            self.assertTrue((bandmem.data() ==
```
I came across `numpy.testing.assert_array_equal`. Maybe it would be worth using that.

timgraham commented:
Could you please add some versionchanged and/or versionadded annotations to gdal.txt to describe what's changed in 1.9?

yellowcap commented:
All comments are integrated. Fixed typos/wording, simplified test assertions, and I introduced versionadded/versionchanged annotations.

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -47,19 +61,67 @@ def __init__(self, ds_input, write=False):
+                self.skew.x, self.skew.y = ds_input['skew']
         else:
-            raise GDALException('Invalid data source input type: "{}".'.format(type(ds_input)))
+            raise GDALException(
```
single line?

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1101,35 +1104,107 @@ one or more layers of data named bands.  Each band, represented by a
+            >>> rst.srs = 4326
+            >>> rst.srs.srid
+            4326
+
```
chop extra newline

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1101,35 +1104,107 @@ one or more layers of data named bands.  Each band, represented by a
+
+.. class:: GDALRaster(ds_input, write=False)
+
+    .. versionchanged:: 1.9
```
I generally prefer putting the "changed" annotations at the bottom of the section (above `.. attribute:: name` in this case)

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1101,35 +1104,107 @@ one or more layers of data named bands.  Each band, represented by a
-        :class:`SpatialReference` instance.
+        The spatial reference system of the `source`, as a
+        :class:`SpatialReference` instance. The SRS can be changed by
+        setting it to an other `SpatialReference` or providing any input
```
need double backticks for it to have a visible effect

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1144,34 +1219,76 @@ blue.
+            2
+            >>> rst.bands[1].data()
+            array([[ 2.,  3.]], dtype=float32)
+
```
chop extra newline

yellowcap replied:
This is not picked up by the github inteface here, but line is removed.

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1216,6 +1343,49 @@ blue.
+        data array if no parameters are provided. A subset of the pixel array can
+        be requested by specifying an offset and block size as tuples.
+
+        If NumPy is available, the data is returned as NumPy array. For performance reasons, it is highly recommended to use NumPy.
```
missing linebreak

timgraham commented:
LGTM docs-wise.

claudep commented:
Pushed in f269c1d6f6dcc, \o/

timgraham commented:
One test failure on Ubuntu 12.04 (Geos 3.2.2):
```
gis_tests.gdal_tests.test_raster.GDALBandTests.test_band_data_setters

Traceback (most recent call last):
  File "/mnt/jenkinsdata/workspace/django-master/database/mysql_gis/python/python3.4/tests/gis_tests/gdal_tests/test_raster.py", line 193, in test_band_data_setters
    'nr_of_bands': 1
  File "/mnt/jenkinsdata/workspace/django-master/database/mysql_gis/python/python3.4/django/contrib/gis/gdal/raster/source.py", line 111, in __init__
    self.srs = ds_input.get('srid', 0)
  File "/mnt/jenkinsdata/workspace/django-master/database/mysql_gis/python/python3.4/django/contrib/gis/gdal/raster/source.py", line 203, in srs
    srs = SpatialReference(value)
  File "/mnt/jenkinsdata/workspace/django-master/database/mysql_gis/python/python3.4/django/contrib/gis/gdal/srs.py", line 95, in __init__
    self.import_epsg(srs_input)
  File "/mnt/jenkinsdata/workspace/django-master/database/mysql_gis/python/python3.4/django/contrib/gis/gdal/srs.py", line 286, in import_epsg
    capi.from_epsg(self.ptr, epsg)
  File "/mnt/jenkinsdata/workspace/django-master/database/mysql_gis/python/python3.4/django/contrib/gis/gdal/prototypes/errcheck.py", line 119, in check_errcode
    check_err(result, cpl=cpl)
  File "/mnt/jenkinsdata/workspace/django-master/database/mysql_gis/python/python3.4/django/contrib/gis/gdal/error.py", line 72, in check_err
    raise e(msg)
django.contrib.gis.gdal.error.SRSException: Unsupported SRS.	
```

claudep commented:
I guess we could set a SRID value to the in-memory raster `rsmem`. Daniel, comments?

yellowcap commented:
I was just looking into this and was able to reproduce the error. It seems that earlier versions of gdal do not like the `0` srid. So I guess we should require an srid to be specified when instantiating from raw data, i.e. replace `self.srs = ds_input.get('srid', 0)` with `self.srs = ds_input.get('srid')`, and update the test data acordingly. The examples in the docs would also be changed to include an srid argument.

yellowcap commented:
Here is a patch that should resolve the problem by making the ``srid`` a requied parameter for ``GDALRaster`` instantiation.

https://github.com/geodesign/django/commit/f26e790fc84ee8d7c8e9e36b924f56041c132585

claudep commented:
Nice, please make it a pull request.

yellowcap commented:
Here it is https://github.com/django/django/pull/4338
